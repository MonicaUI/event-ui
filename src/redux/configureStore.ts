import analyticsConfig from './analytics';
import reducer from './reducer';

// using require instead of import for conditional compilation;
let createStoreWithMiddleware;
if (process.env.NODE_ENV === 'production') {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  createStoreWithMiddleware = require('nucleus-guestside-site/src/redux/prodCreateStoreWithMiddleware').default;
} else {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  createStoreWithMiddleware = require('nucleus-guestside-site/src/redux/devCreateStoreWithMiddleware').default;
}

export default function configureStore(
  initialState: $TSFixMe,
  baseAnalyticsConfig = {},
  thunkExtraArgument?: $TSFixMe
): $TSFixMe {
  return createStoreWithMiddleware(reducer, initialState, {
    analyticsConfig: analyticsConfig(initialState, baseAnalyticsConfig),
    thunkExtraArgument
  });
}
