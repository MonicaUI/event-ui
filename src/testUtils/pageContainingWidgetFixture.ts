export default function pageContainingWidgetFixture(pageId: $TSFixMe, widgetId: $TSFixMe): $TSFixMe {
  const pages = {
    [pageId]: {
      id: pageId,
      name: pageId,
      title: pageId,
      version: 1,
      rootLayoutItemIds: ['container'],
      type: 'PAGE'
    }
  };
  const layoutItems = {
    container: {
      id: 'container',
      layout: {
        type: 'container',
        parentId: null,
        childIds: [widgetId]
      }
    },
    [widgetId]: {
      id: widgetId,
      layout: {
        type: 'page',
        parentId: 'container',
        childIds: []
      },
      config: {}
    }
  };
  return { pages, layoutItems };
}
