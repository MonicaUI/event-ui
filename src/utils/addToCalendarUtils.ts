import { createSelector } from 'reselect';
import CalendarDescriptionSource from '../dialogs/DownloadCalendarDialog/CalendarDescriptionSource';
import moment from 'moment';
import { getIn } from 'icepick';
import qs from 'querystring';
import { CalendarType } from '../dialogs/DownloadCalendarDialog/CalendarDownloadFileType';

const TITLE = '<<TITLE>>';
const STARTDATE = '<<STARTDATE>>';
const ENDDATE = '<<ENDDATE>>';
const LOCATION = '<<LOCATION>>';
const DESCRIPTION = '<<DESCRIPTION>>';
const EMPTY_STUB = '00000000-0000-0000-0000-000000000000';
/*
 * We are not saving the actual eventId for mapping in redux state as eventId is unique to any event
 * But for sessions, we are maintaining the mapping between actual sessionId and sessionCalendarUrl
 */
export const CALENDAR_EVENT_ID = EMPTY_STUB;

export const AddToCalendarExperiment = {
  ICS_CALENDAR_VARIANT: 0,
  GOOGLE_CALENDAR_FOR_INPERSON_VARIANT: 1,
  GOOGLE_CALENDAR_FOR_VIRTUAL_VARIANT: 2,
  OUTLOOK_ICS_ICAL_CALENDAR_VARIANT: 3
};

const formatDate = date => {
  let formattedDate = `${moment.utc(date).format('YYYY-MM-DDTHH:mm:ss')}Z`;
  formattedDate = formattedDate.replace(/-/g, '');
  return formattedDate.replace(/:/g, '');
};

/**
 * This method creates url for ics calendar when called using calendarService
 */
const calendarServicePathForIcs = (calendarBasePath, eventId, sessionId, queryParamString, calendarType) => {
  const extendedPath = `${calendarBasePath}calendar/v1/calendars/${calendarType}/events/${eventId}`;

  return sessionId
    ? `${extendedPath}/sessions/${sessionId}?${queryParamString}`
    : `${extendedPath}?${queryParamString}`;
};

/**
 * The method created url for ics calendar when called using guestSideService
 */
const guestSideServicePathForIcs = (calendarBasePath, eventId, sessionId, queryParamString) => {
  const extendedPath = `${calendarBasePath}event/ical/${eventId}`;

  return sessionId ? `${extendedPath}/${sessionId}?${queryParamString}` : `${extendedPath}?${queryParamString}`;
};

export function formatDownloadIcsFileUrl({
  calendarBasePath,
  environment,
  locale,
  sessionId,
  previewToken,
  testModeHash,
  eventId,
  addToCalendarEnabled,
  calendarType,
  attendeeId
}: {
  calendarBasePath?: $TSFixMe;
  environment?: $TSFixMe;
  locale?: $TSFixMe;
  sessionId?: $TSFixMe;
  previewToken?: $TSFixMe;
  testModeHash?: $TSFixMe;
  eventId?: $TSFixMe;
  addToCalendarEnabled?: $TSFixMe;
  calendarType?: $TSFixMe;
  attendeeId?: $TSFixMe;
}): $TSFixMe {
  const queryParams = {};

  if (environment) {
    (queryParams as $TSFixMe).environment = environment;
  }

  if (locale) {
    (queryParams as $TSFixMe).locale = locale;
  }

  if (previewToken) {
    (queryParams as $TSFixMe).previewToken = previewToken;
  } else if (testModeHash) {
    (queryParams as $TSFixMe).tm = testModeHash;
  }

  if (attendeeId) {
    (queryParams as $TSFixMe).inviteeId = attendeeId;
  }

  const queryParamString = qs.stringify(queryParams);
  return addToCalendarEnabled
    ? calendarServicePathForIcs(calendarBasePath, eventId, sessionId, queryParamString, calendarType)
    : guestSideServicePathForIcs(calendarBasePath, eventId, sessionId, queryParamString);
}

export function downloadEventIcsFile(
  calendarBasePath: $TSFixMe,
  environment: $TSFixMe,
  previewToken: $TSFixMe,
  testModeHash: $TSFixMe,
  eventId: $TSFixMe,
  addToCalendarEnabled: $TSFixMe,
  calendarType: $TSFixMe,
  attendeeId: $TSFixMe,
  locale: $TSFixMe
): $TSFixMe {
  const calendarPathUrl = formatDownloadIcsFileUrl({
    calendarBasePath,
    environment,
    locale,
    previewToken,
    testModeHash,
    eventId,
    addToCalendarEnabled,
    calendarType,
    attendeeId
  });
  global.location.replace(calendarPathUrl);
}

export function getCalendarUrl(
  calendarProviderURL: $TSFixMe,
  eventOrSessionInfo: $TSFixMe,
  isEventCalendar = false,
  calendarDescription = {}
): $TSFixMe {
  const startDate = formatDate(isEventCalendar ? eventOrSessionInfo.startDate : eventOrSessionInfo.startTime);
  const endDate = formatDate(isEventCalendar ? eventOrSessionInfo.endDate : eventOrSessionInfo.endTime);
  let calendarUrl = calendarProviderURL.replace(TITLE, encodeURIComponent(eventOrSessionInfo.title));
  calendarUrl = calendarUrl.replace(STARTDATE, startDate);
  calendarUrl = calendarUrl.replace(ENDDATE, endDate);
  calendarUrl = calendarUrl.replace(LOCATION, encodeURIComponent(eventOrSessionInfo.location));
  if (
    isEventCalendar &&
    calendarDescription &&
    (calendarDescription as $TSFixMe).descriptionSource !== CalendarDescriptionSource.EVENT
  ) {
    calendarUrl =
      (calendarDescription as $TSFixMe).descriptionSource === CalendarDescriptionSource.CUSTOM
        ? calendarUrl.replace(DESCRIPTION, encodeURIComponent((calendarDescription as $TSFixMe).customDescription))
        : calendarUrl.replace(DESCRIPTION, '');
  } else {
    calendarUrl = calendarUrl.replace(DESCRIPTION, encodeURIComponent(eventOrSessionInfo.description));
  }
  return calendarUrl;
}

export const getEventInfo = createSelector(
  state => (state as $TSFixMe).text.translate,
  state => (state as $TSFixMe).event.title,
  state => (state as $TSFixMe).event.startDate,
  state => (state as $TSFixMe).event.endDate,
  state => (state as $TSFixMe).event.location,
  state => (state as $TSFixMe).event.description,
  (translate, title, startDate, endDate, location, description) => {
    return {
      title: translate(title),
      startDate,
      endDate,
      location: translate(location),
      description: translate(description)
    };
  }
);

export const getCalendarSettings = (state: $TSFixMe): $TSFixMe => {
  return getIn(state, ['appData', 'calendarSettings']);
};

export const isMobileView = (): $TSFixMe => {
  const maxMobileViewScreenSize = 480;
  const screenSize = window.innerWidth || document.documentElement.clientWidth || document.body.clientWidth;
  return screenSize < maxMobileViewScreenSize;
};

export function getEventCalendarDescription(calendarSettings: $TSFixMe): $TSFixMe {
  return {
    descriptionSource: calendarSettings.descriptionSource,
    customDescription: calendarSettings.customDescription
  };
}

/**
 * @returns preview token if present to be used in calendar service call
 */
export function getPreviewToken(isPreview: $TSFixMe, pathInfo: $TSFixMe): $TSFixMe {
  return isPreview ? pathInfo?.queryParams?.previewToken : null;
}

/**
 * @returns test mode hash if present to be used in calendar service call
 */
export function getTestModeHash(isTestMode: $TSFixMe, pathInfo: $TSFixMe): $TSFixMe {
  return isTestMode ? pathInfo?.queryParams?.tm : null;
}

/**
 * @returns true when outlook_ics_ical calendar virtual variant is ON for the account
 */
export function isAddToCalendarEnabled(state: $TSFixMe): $TSFixMe {
  return (
    state.experiments?.flexAddToGoogleCalendarExperimentVariant >=
    AddToCalendarExperiment.OUTLOOK_ICS_ICAL_CALENDAR_VARIANT
  );
}

/**
 * @returns true when google calendar virtual variant is ON for the account
 */
export function isFlexAddToGoogleCalendarVirtualEventVariant(state: $TSFixMe): $TSFixMe {
  return (
    state.experiments?.flexAddToGoogleCalendarExperimentVariant >=
    AddToCalendarExperiment.GOOGLE_CALENDAR_FOR_VIRTUAL_VARIANT
  );
}

/**
 * @returns boolean true, when google calendar in person variant is ON for the account
 */
export function isFlexAddToGoogleCalendarInPersonEventVariant(state: $TSFixMe): $TSFixMe {
  return (
    state.experiments?.flexAddToGoogleCalendarExperimentVariant ===
    AddToCalendarExperiment.GOOGLE_CALENDAR_FOR_INPERSON_VARIANT
  );
}

/**
 * @returns boolean true, when google calendar is enabled for that account
 */
export function isFlexAddToGoogleCalendarEnabledVariant(state: $TSFixMe): $TSFixMe {
  return (
    state.experiments?.flexAddToGoogleCalendarExperimentVariant >=
    AddToCalendarExperiment.GOOGLE_CALENDAR_FOR_INPERSON_VARIANT
  );
}

/**
 * @returns if sessionId is present in state, return google session calendar, else return google event calendar present
 * in state otherwise return undefined
 */
export function getGoogleCalendar(calendarState: $TSFixMe, entityId = CALENDAR_EVENT_ID): $TSFixMe {
  if (entityId !== CALENDAR_EVENT_ID) {
    return calendarState?.google?.[entityId];
  }
  return calendarState?.google?.[CALENDAR_EVENT_ID];
}

/**
 * @returns if sessionId is present in state, return google session calendar, else return outlook event calendar present
 * in state otherwise return undefined
 */
export function getOutlookCalendar(calendarState: $TSFixMe, entityId = CALENDAR_EVENT_ID): $TSFixMe {
  if (entityId !== CALENDAR_EVENT_ID) {
    return calendarState?.outlook?.[entityId];
  }
  return calendarState?.outlook?.[CALENDAR_EVENT_ID];
}

export function openCalendar(fileType: $TSFixMe, entityId: $TSFixMe, calendarProviders: $TSFixMe): $TSFixMe {
  if (fileType === CalendarType.GOOGLE) {
    return getGoogleCalendar(calendarProviders, entityId);
  } else if (fileType === CalendarType.OUTLOOK) {
    return getOutlookCalendar(calendarProviders, entityId);
  }
}
