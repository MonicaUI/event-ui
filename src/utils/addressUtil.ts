/**
 * Compares both city level address and returns true if there is any diff
 * @param {Object} cityLevelAddress
 * @param {Object} cityLevelAddressFromAttendee
 * @returns {boolean}
 */
export function hasAddressChangedInState(cityLevelAddress: $TSFixMe, cityLevelAddressFromAttendee: $TSFixMe): $TSFixMe {
  return (
    cityLevelAddress &&
    cityLevelAddressFromAttendee &&
    (cityLevelAddress.city !== cityLevelAddressFromAttendee.city ||
      cityLevelAddress.stateCode !== cityLevelAddressFromAttendee.stateCode ||
      cityLevelAddress.stateName !== cityLevelAddressFromAttendee.stateName ||
      cityLevelAddress.postalCode !== cityLevelAddressFromAttendee.postalCode ||
      cityLevelAddress.countryCode !== cityLevelAddressFromAttendee.countryCode ||
      cityLevelAddress.countryName !== cityLevelAddressFromAttendee.countryName)
  );
}

/**
 * Get address upto city from attendee address
 * @param {Object} attendeeAddress
 * @returns {Object}
 */
export function getCityLevelAddressFromAttendee(attendeeAddress: $TSFixMe): $TSFixMe {
  return {
    city: attendeeAddress.city,
    stateCode: attendeeAddress.stateCode,
    stateName: attendeeAddress.state,
    postalCode: attendeeAddress.postalCode,
    countryCode: attendeeAddress.countryCode,
    countryName: attendeeAddress.country
  };
}
