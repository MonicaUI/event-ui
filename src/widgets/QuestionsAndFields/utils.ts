import { getIn } from 'icepick';
import getJSONValue from 'nucleus-widgets/utils/fields/getJSONValue';
import { getContactFieldForWidget } from 'event-widgets/redux/selectors/website/pluginData/registrationProcessNavigation';
import { getSubQuestionParentQuestionIdJsonPath } from 'event-widgets/redux/selectors/appData';
import { NAChoice, OtherChoice } from '@cvent/nucleus-form-validations';

const DEFAULT_QUESTION_CHOICES = Object.freeze([]);

export function getConfigWithDisplay(config: $TSFixMe, display: $TSFixMe): $TSFixMe {
  return {
    ...config,
    appData: {
      ...config.appData,
      display
    }
  };
}

export function convertChoicesToBaseLanguage(
  choices: $TSFixMe,
  choicesInCurrentLanguage: $TSFixMe,
  choicesInBaseLanguage: $TSFixMe
): $TSFixMe {
  return choices
    .map(choice => {
      const localizedChoice = choicesInCurrentLanguage.find(c => c.text === choice);
      if (localizedChoice) {
        const baseLangChoice = choicesInBaseLanguage.find(c => c.id === localizedChoice.id);
        if (baseLangChoice) {
          return {
            choice: baseLangChoice.text,
            answerType: 'Choice'
          };
        }
      } else {
        return {
          choice,
          answerType: 'Choice'
        };
      }
    })
    .filter(Boolean);
}

const emptyOrNaOrOther = value => !value || value === NAChoice || value === OtherChoice;

const needsTranslation = (baseChoices, localizedChoices, text) =>
  !localizedChoices.find(c => c.text === text) && baseChoices.find(c => c.text === text);

export function setSelectedValuesForChoices(
  answer: $TSFixMe,
  baseChoices: $TSFixMe,
  localizedChoices: $TSFixMe
): $TSFixMe {
  const translateEachIfNecessary = selectedValues =>
    selectedValues
      .filter(value => !emptyOrNaOrOther(value))
      .map(text =>
        needsTranslation(baseChoices, localizedChoices, text)
          ? updateSelectedValues(text, baseChoices, localizedChoices)
          : text
      );
  const selectedValues = translateEachIfNecessary(answer.value.selectedValues || []);
  const selectedPrevValues = translateEachIfNecessary(getIn(answer, ['valueBeforeMod', 'selectedValues']) || []);

  addMissingChoicesToLocaledAnswer(selectedValues, selectedPrevValues, answer);

  return {
    ...answer,
    value: {
      ...answer.value,
      selectedValues: selectedValues.length ? selectedValues : answer.value.selectedValues || []
    },
    valueBeforeMod: {
      ...answer.valueBeforeMod,
      selectedValues: selectedPrevValues.length
        ? selectedPrevValues
        : getIn(answer, ['valueBeforeMod', 'selectedValues']) || []
    }
  };
}

function addOtherOrNAChoiceToLocalizedAnswer(selectedValues, answerSelectedValues, otherOrNAChoice) {
  if (selectedValues.length && answerSelectedValues.includes(otherOrNAChoice)) {
    selectedValues.push(otherOrNAChoice);
  }
}

function addMissingChoicesToLocaledAnswer(selectedValues, selectedPrevValues, answer) {
  const answerSelectedValues = getIn(answer, ['value', 'selectedValues']) || [];
  const answerSelectedPrevValues = getIn(answer, ['valueBeforeMod', 'selectedValues']) || [];
  addOtherOrNAChoiceToLocalizedAnswer(selectedValues, answerSelectedValues, NAChoice);
  addOtherOrNAChoiceToLocalizedAnswer(selectedPrevValues, answerSelectedPrevValues, NAChoice);
  addOtherOrNAChoiceToLocalizedAnswer(selectedValues, answerSelectedValues, OtherChoice);
  addOtherOrNAChoiceToLocalizedAnswer(selectedPrevValues, answerSelectedPrevValues, OtherChoice);
}

function updateSelectedValues(choiceText, baseChoices, localizedChoices) {
  const choice = baseChoices.find(c => c.text === choiceText);
  if (choice) {
    const updatedChoice = localizedChoices.find(c => c.id === choice.id);
    if (updatedChoice) {
      return updatedChoice.text;
    }
  }
}

export function getLocalizedContactFieldForWidget(state: $TSFixMe, config: $TSFixMe, id: $TSFixMe): $TSFixMe {
  const registrationField = getContactFieldForWidget(state, config.registrationFieldPageType, config.fieldId, id);
  return {
    ...registrationField,
    displayName: getIn(config, ['appData', 'displayName']) || registrationField.displayName
  };
}

export function modifyFieldDefinitionForMultiLanguage(fieldDefinition: $TSFixMe): $TSFixMe {
  const modifiedFieldDefinition = {
    ...fieldDefinition,
    question: {
      ...fieldDefinition.question,
      questionTypeInfo: {
        ...fieldDefinition.question.questionTypeInfo,
        choices: fieldDefinition.question.questionTypeInfo.choices.map(choice => {
          return { id: choice.text.slice(choice.text.indexOf('.') + 1), text: choice.text };
        })
      }
    }
  };
  return modifiedFieldDefinition;
}

export function setParentQuestionId(appData: $TSFixMe, config: $TSFixMe, questionId: $TSFixMe): $TSFixMe {
  return {
    ...config,
    appData: {
      ...config.appData,
      parentQuestionId: getJSONValue(appData, getSubQuestionParentQuestionIdJsonPath(appData, questionId))
    }
  };
}

export function getQuestionCategory(question: $TSFixMe): $TSFixMe {
  if (question.isProductQuestion) {
    return 'productQuestions';
  }
  if (question.isTravelQuestion) {
    return 'travelQuestions';
  }
  return 'registrationQuestions';
}

export function setDefaultChoicesPreSelected(
  appData: $TSFixMe,
  questionCategory: $TSFixMe,
  questionId: $TSFixMe,
  answer: $TSFixMe,
  baseChoices: $TSFixMe
): $TSFixMe {
  const questionDefaultChoices = getIn(
    appData,
    ['registrationSettings', questionCategory, questionId, 'question', 'questionTypeInfo', 'defaultChoice'],
    DEFAULT_QUESTION_CHOICES
  );
  let filteredChoices = DEFAULT_QUESTION_CHOICES;
  if (questionDefaultChoices && questionDefaultChoices.length > 0) {
    if (!Array.isArray(questionDefaultChoices)) {
      filteredChoices = baseChoices.filter(c => c.id === questionDefaultChoices).map(choice => choice.text);
    } else {
      filteredChoices = baseChoices.map(choice => {
        const matchIndex = questionDefaultChoices.findIndex(defaultChoice => defaultChoice.id === choice.id);
        if (matchIndex !== -1) {
          return choice.text;
        }
      });
    }
    filteredChoices = filteredChoices.filter(x => x !== undefined);
    return {
      ...answer,
      value: {
        ...answer.value,
        selectedValues: filteredChoices
      },
      valueBeforeMod: {
        ...answer.valueBeforeMod,
        selectedValues: filteredChoices
      }
    };
  }
  return {
    ...answer,
    value: {
      ...answer.value,
      useDefaultChoices: false // answer = undefined and Question does not have defautl choices
    }
  };
}
