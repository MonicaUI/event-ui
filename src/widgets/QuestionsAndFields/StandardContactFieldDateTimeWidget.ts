import { connect } from 'react-redux';
import StandardContactFields from 'event-widgets/lib/StandardContactFields/StandardContactFields';
import StandardContactFieldDateTimeWidget from 'event-widgets/lib/StandardContactFields/StandardContactFieldDateTimeWidget';
import { resolveDateFromString, getLocaleDateString } from '@cvent/nucleus-core-datetime-utils';
import { getLocalizedContactFieldForWidget } from './utils';
import * as eventRegistrationData from '../../utils/eventRegistrationData';
import {
  isGroupRegistration,
  getEventRegistrationId,
  isGroupLeader,
  isViewingGuest
} from '../../redux/selectors/currentRegistrant';
import { evaluateQuestionVisibilityLogic } from '../../redux/actions';
import { DATE_CHANGE_EMPTY_DATE, DATE_CHANGE_SUCCESS } from 'nucleus-core/forms/PickADateBase';
import withMappedWidgetConfig from 'nucleus-widgets/lib/withMappedWidgetConfig';
import withMemoizedFunctions from 'nucleus-widgets/lib/withMemoizedFunctions';

const getUTCTime = date => date.getTime() - date.getTimezoneOffset() * 60000;

const setAnswer = (baseAction, value, locale) => {
  const { dateText } = value;
  const format = getLocaleDateString(locale);
  const date = dateText && resolveDateFromString(dateText, format);
  return eventRegistrationData.setAnswerAction(baseAction, date && getUTCTime(date));
};

export default withMappedWidgetConfig(
  connect(
    withMemoizedFunctions({ getLocalizedContactFieldForWidget })(memoized => (state: $TSFixMe, props: $TSFixMe) => {
      const registrationField = memoized.getLocalizedContactFieldForWidget(state, props.config, props.id);
      const fieldConfig = StandardContactFields[props.config.fieldId];

      const eventRegistrationPath = eventRegistrationData.buildEventRegistrationPath(fieldConfig.regApiPath);

      const alwaysEnforceRequiredValidation = fieldConfig.alwaysEnforceRequiredValidation;
      const currentEventRegistrationId = getEventRegistrationId(state);

      const isPrimaryRegistrant = isGroupLeader(state, currentEventRegistrationId) || !isGroupRegistration(state);
      const shouldDisableFieldForPlannerReg =
        state.defaultUserSession.isPlanner && fieldConfig.plannerRegistrationReadOnly;

      const answer = eventRegistrationData.answer({
        state,
        widgetConfig: props.config,
        eventRegistrationPath,
        // converting epoch time to ISO format.
        getAnswerFormatter: value => (value ? new Date(value).toISOString() : null)
      });

      return {
        registrationField,
        eventRegistrationId: getEventRegistrationId(state),
        skipRequiredValidation: state.defaultUserSession.isPlanner && !alwaysEnforceRequiredValidation,
        plannerRegistrationReadOnly: isPrimaryRegistrant ? shouldDisableFieldForPlannerReg : false,
        value: answer.value,
        setterAction: answer.setterAction,
        locale: state.text.locale,
        isViewingGuest: isViewingGuest(state)
      };
    }),
    { setAnswer, evaluateQuestionVisibilityLogic },
    (stateProps: $TSFixMe, dispatchProps: $TSFixMe, ownProps: $TSFixMe) => {
      return {
        ...ownProps,
        ...stateProps,
        ...dispatchProps,
        onDateTimeChange: value => {
          dispatchProps.setAnswer(stateProps.setterAction, value, stateProps.locale);
          // Only evaluate visibility logic if the date is valid, or the field is empty.
          if (value.statusCode === DATE_CHANGE_SUCCESS || value.statusCode === DATE_CHANGE_EMPTY_DATE) {
            dispatchProps.evaluateQuestionVisibilityLogic(ownProps.config.fieldId, false, stateProps.isViewingGuest);
          }
        }
      };
    }
  )(StandardContactFieldDateTimeWidget)
);
