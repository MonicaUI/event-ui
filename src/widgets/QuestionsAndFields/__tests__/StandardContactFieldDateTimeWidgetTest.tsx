import React from 'react';
import StandardContactFieldDateTimeWidget from '../StandardContactFieldDateTimeWidget';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from '../../../redux/configureStore';
import WidgetFactory from '../../../widgetFactory';
import Grid from 'nucleus-core/layout/flexbox/Grid';
import { NOT_VISIBLE } from 'cvent-question-widgets/lib/DisplayType';
import * as registrationFieldPageType from 'event-widgets/utils/registrationFieldPageType';
import pageContainingWidgetFixture from '../../../testUtils/pageContainingWidgetFixture';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const initialState = {
  website: {
    ...pageContainingWidgetFixture('pageId', 'widgetId'),
    pluginData: {
      registrationProcessNavigation: {
        registrationPaths: {
          regPathId: {
            id: 'regPathId',
            pageIds: ['pageId']
          }
        }
      }
    },
    siteInfo: {
      sharedConfigs: {}
    }
  },
  widgetFactory: new WidgetFactory(),
  appData: {
    registrationSettings: {
      registrationPaths: {
        regPathId: {
          registrationPageFields: {
            [registrationFieldPageType.Registration]: {
              registrationFieldPageType: registrationFieldPageType.Registration,
              registrationFields: {
                'b468074b-b6a7-4149-8f97-e6b21b2ca8b3': {
                  fieldName: 'Date of Birth',
                  fieldId: 'b468074b-b6a7-4149-8f97-e6b21b2ca8b3',
                  displayName: 'Date of Birth',
                  display: NOT_VISIBLE,
                  isCustomField: false
                }
              }
            }
          }
        }
      }
    }
  },
  registrationForm: {
    currentEventRegistrationId: 'eventRegistration1',
    regCart: {
      eventRegistrations: {
        eventRegistration1: {
          registrationPathId: 'regPathId'
        }
      }
    }
  },
  text: {
    locale: 'en-us'
  },
  userSession: {},
  defaultUserSession: {},
  clients: {},
  event: {}
};

const defaultProps = {
  translate: (resx, opts) => (opts ? `${resx}:${JSON.stringify(opts)}` : resx),
  style: {},
  classes: {},
  id: 'widgetId'
};

test('datetime standard contact field renders and allows you to enter datetime', () => {
  const store = configureStore(initialState);
  const component = mount(
    <Provider store={store}>
      <Grid>
        <StandardContactFieldDateTimeWidget
          {...defaultProps}
          config={{
            fieldId: 'b468074b-b6a7-4149-8f97-e6b21b2ca8b3',
            registrationFieldPageType: registrationFieldPageType.Registration
          }}
          type="EventStandardContactFieldDateTime"
        />
      </Grid>
    </Provider>
  );

  // unanswered view
  expect(component).toMatchSnapshot();

  // enter date
  component
    .find('[data-cvent-id="input"]')
    .hostNodes()
    .simulate('change', { target: { value: '11/02/2006' } });
  expect(component).toMatchSnapshot();
  expect(store.getState().registrationForm.regCart.eventRegistrations.eventRegistration1.attendee).toMatchSnapshot();

  // enter text in date
  component
    .find('[data-cvent-id="input"]')
    .hostNodes()
    .simulate('change', { target: { value: '3' } });
  expect(store.getState().registrationForm.regCart.eventRegistrations.eventRegistration1.attendee).toMatchSnapshot();
});

describe('Visibility logic tests', () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore(initialState);
    component = mount(
      <Provider store={store}>
        <Grid>
          <StandardContactFieldDateTimeWidget
            {...defaultProps}
            config={{
              fieldId: 'b468074b-b6a7-4149-8f97-e6b21b2ca8b3',
              registrationFieldPageType: registrationFieldPageType.Registration
            }}
            type="EventStandardContactFieldDateTime"
          />
        </Grid>
      </Provider>
    );
  });
  it('Should evaluate visibility logic when a valid date is input', () => {
    component
      .find('[data-cvent-id="input"]')
      .hostNodes()
      .simulate('change', { target: { value: '10/16/2019' } });
    expect(store.getActions()).toMatchSnapshot();
  });
  it('Should evaluate visibilty logic when a empty value is input', () => {
    component
      .find('[data-cvent-id="input"]')
      .hostNodes()
      .simulate('change', { target: { value: '' } });
    expect(store.getActions()).toMatchSnapshot();
  });
  it('Should not evaluate visibility logic when an invalid date is input', () => {
    component
      .find('[data-cvent-id="input"]')
      .hostNodes()
      .simulate('change', { target: { value: '1/2' } });
    expect(store.getActions()).toMatchSnapshot();
  });
});
