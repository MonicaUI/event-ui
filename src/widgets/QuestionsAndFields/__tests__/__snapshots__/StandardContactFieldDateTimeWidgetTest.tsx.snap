// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visibility logic tests Should evaluate visibility logic when a valid date is input 1`] = `
Array [
  Object {
    "payload": Object {
      "path": Array [
        "eventRegistrations",
        "eventRegistration1",
        "attendee",
        "personalInformation",
        "dateOfBirth",
      ],
      "value": 1571184000000,
    },
    "type": "event-guestside-site/regCart/SET_REG_CART_FIELD_VALUE",
  },
  Object {
    "payload": Object {
      "questionIds": Array [],
      "visibility": Object {},
    },
    "type": "event-guestside-site/UPDATE_QUESTION_VISIBILITY",
  },
]
`;

exports[`Visibility logic tests Should evaluate visibilty logic when a empty value is input 1`] = `
Array [
  Object {
    "payload": Object {
      "path": Array [
        "eventRegistrations",
        "eventRegistration1",
        "attendee",
        "personalInformation",
        "dateOfBirth",
      ],
      "value": "",
    },
    "type": "event-guestside-site/regCart/SET_REG_CART_FIELD_VALUE",
  },
  Object {
    "payload": Object {
      "questionIds": Array [],
      "visibility": Object {},
    },
    "type": "event-guestside-site/UPDATE_QUESTION_VISIBILITY",
  },
]
`;

exports[`Visibility logic tests Should not evaluate visibility logic when an invalid date is input 1`] = `
Array [
  Object {
    "payload": Object {
      "path": Array [
        "eventRegistrations",
        "eventRegistration1",
        "attendee",
        "personalInformation",
        "dateOfBirth",
      ],
      "value": null,
    },
    "type": "event-guestside-site/regCart/SET_REG_CART_FIELD_VALUE",
  },
]
`;

exports[`datetime standard contact field renders and allows you to enter datetime 1`] = `
<div
  class=""
>
  <div
    class=""
    style="width: 100%;"
  >
    <div
      class=""
    >
      <div
        class=""
        style="flex: 1 1 auto;"
      >
        <div
          class=""
        >
          <label
            class="labelOnTop"
            for="b468074b-b6a7-4149-8f97-e6b21b2ca8b3"
          >
            <span
              class=""
            >
              <span
                data-cvent-id="label"
              >
                <span>
                  Date of Birth
                </span>
              </span>
            </span>
          </label>
          <div
            class=""
          >
            <div
              class=""
              data-cvent-id="trigger"
            >
              <span>
                <div
                  class=""
                >
                  <input
                    autocomplete="off"
                    class="textboxDate"
                    data-cvent-id="input"
                    dateinputlabel="NucleusWidgets_DateTimeQuestion_DateInputLabel__resx"
                    id="b468074b-b6a7-4149-8f97-e6b21b2ca8b3"
                    name="b468074b-b6a7-4149-8f97-e6b21b2ca8b3"
                    periodconfig="[object Object]"
                    placeholder="NucleusCoreComponent_PickADate_TextboxPlaceholder_mm/dd/yyyy__resx"
                    separatorsymbol=":"
                    timeinputlabel="NucleusWidgets_DateTimeQuestion_TimeInputLabel__resx"
                    type="text"
                    value=""
                  />
                </div>
              </span>
              <span
                class="mockNucleusCoreTransition"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`datetime standard contact field renders and allows you to enter datetime 2`] = `
<div
  class=""
>
  <div
    class=""
    style="width: 100%;"
  >
    <div
      class=""
    >
      <div
        class=""
        style="flex: 1 1 auto;"
      >
        <div
          class=""
        >
          <label
            class="labelOnTop"
            for="b468074b-b6a7-4149-8f97-e6b21b2ca8b3"
          >
            <span
              class=""
            >
              <span
                data-cvent-id="label"
              >
                <span>
                  Date of Birth
                </span>
              </span>
            </span>
          </label>
          <div
            class=""
          >
            <div
              class=""
              data-cvent-id="trigger"
            >
              <span>
                <div
                  class=""
                >
                  <input
                    autocomplete="off"
                    class="textboxDate"
                    data-cvent-id="input"
                    dateinputlabel="NucleusWidgets_DateTimeQuestion_DateInputLabel__resx"
                    id="b468074b-b6a7-4149-8f97-e6b21b2ca8b3"
                    name="b468074b-b6a7-4149-8f97-e6b21b2ca8b3"
                    periodconfig="[object Object]"
                    placeholder="NucleusCoreComponent_PickADate_TextboxPlaceholder_mm/dd/yyyy__resx"
                    separatorsymbol=":"
                    timeinputlabel="NucleusWidgets_DateTimeQuestion_TimeInputLabel__resx"
                    type="text"
                    value="11/02/2006"
                  />
                </div>
              </span>
              <span
                class="mockNucleusCoreTransition"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`datetime standard contact field renders and allows you to enter datetime 3`] = `
Object {
  "personalInformation": Object {
    "dateOfBirth": 1162425600000,
  },
}
`;

exports[`datetime standard contact field renders and allows you to enter datetime 4`] = `
Object {
  "personalInformation": Object {
    "dateOfBirth": null,
  },
}
`;
